### PR TITLE
freecad: make customizable

### PR DESCRIPTION
**Pull request by ein-shved**
FreeCad has addon system with various addons around the net.

It has an addon-manager which allows to browse through registered addons and install them in runtime. But this is not nix-way, because you have to install addons again after system configuration moving.

Additionally freecad allows you to manually install addons and put them to common folder or specify with command arguments.

This patch introduces extra `customize` method to FreeCad derivation attrset (inspired from vim) which allows you to inject addons from nix configuration.

```nix
{ freecad
, fetchFromGitHub
}:
let
  cad-exchanger = fetchFromGitHub {
    owner = "yorikvanhavre";
    repo = "CADExchanger";
    rev = "5c2cd792ddc4581b917ebe7add5ef960bf6c3e2a";
    hash = "sha256-AST5bwhgMbvW3m8V1cv5PqKjJi2eSE1lbXpVLvRVzM8=";
  };
  open-theme = fetchFromGitHub {
    owner = "obelisk79";
    repo = "OpenTheme";
    rev = "80399168ab27c660a9a14812b67492f412f3148f";
    hash = "sha256-NgrBvP+wRNED3JI8zG9JMT//xfZfivQC8yy4m2GsujM=";
  };

  freecad-customized = freecad.customize {
    modules = [ cad-exchanger open-theme ];
    pythons = [
      (ps: with ps; [ requests pyjwt tzlocal ])
    ];
    makeWrapperFlags = [ "--set-default" "QT_FONT_DPI" "85" ];
    userCfg = ./my-default-config.cfg;
  };
in
freecad-customized
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ein-shved** [comment]
@pinpox Hi, I've saw somewhere a file with list of modules for freecad which are installed by default by ondsel. But I missed it and can not find again. Do you have a full list of modules from ondsel to test this with?

**Comment by Pandapip1** [comment]
I can't seem to understand what this code is doing. Would you mind adding documentation for how to customize freecad?

Additionally, I'd argue this merits a changelog entry.

**Comment by ein-shved** [comment]
@Pandapip1 sorry, but where do you expect to see documentation additionally? Inline in code? 

And can you please link me a documentation how to add a changelog entry?

**Comment by Pandapip1** [comment]
> Where do you expect to see documentation additionally? Inline in code?

Oh duh, you can't add stuff to the wiki in a PR. Please disregard, although if/when this is merged, I highly recommend you create a page at https://wiki.nixos.org/wiki/FreeCAD.

> And can you please link me a documentation how to add a changelog entry?

You edit `nixos/doc/manual/release-notes/rl-2411.section.md` to add a quick bullet point noting the change you made to the package.

**Comment by ein-shved** [comment]
> I highly recommend you create a page at https://wiki.nixos.org/wiki/FreeCAD

Of course I'will do.



> You edit `nixos/doc/manual/release-notes/rl-2411.section.md` to add a quick bullet point noting the change you made to the package

Thnx!

File: pkgs/by-name/fr/freecad/freecad-utils.nix [hunk 1]
```diff
@@ -0,0 +1,57 @@
+{
+  runCommand,
+  buildEnv,
+  makeWrapper,
+  lib,
+}:
+rec {
```
Comment (AndersonTorres) [inline pkgs/by-name/fr/freecad/freecad-utils.nix:7]:
```suggestion
{
```

Remove `rec`.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/fr/freecad/freecad-utils.nix (hunk 1 above)
Comment (ein-shved) [reply pkgs/by-name/fr/freecad/freecad-utils.nix:9]: No, the `makeCustomizable` function used inside its own (`override`, `overrideAttrs` and `customize` methods of its result)

Comment (Pandapip1) [reply pkgs/by-name/fr/freecad/freecad-utils.nix:9]: Couldn't this be a recursive variable in a `let`?

Comment (ein-shved) [reply pkgs/by-name/fr/freecad/freecad-utils.nix:9]: Maybe. Gonna try. But can you explain why `let-in` is better? Did not hear that `rec` is anti-pattern.

Comment (ein-shved) [reply pkgs/by-name/fr/freecad/freecad-utils.nix]: Done

**Comment by ein-shved** [comment]
@kirillrdy @Artturin Hi! Sorry for interrupting, but you have made submitting to freecad package. Do I have to do something here?

**Comment by kirillrdy** [comment]
@ein-shved i don't have any strong options about freecad, current listed maintainers are cc @gebner  @AndersonTorres @srounce

File: nixos/doc/manual/release-notes/rl-2411.section.md [hunk 2]
```diff
@@ -858,6 +858,8 @@
 
 - `qgis` and `qgis-ltr` are now built without `grass` by default. `grass` support can be enabled with `qgis.override { withGrass = true; }`.
 
+- `freecad` is now customizable package, which allows to call `customize` and save its configuration and modules in nix-way.
```
Comment (AndersonTorres) [inline nixos/doc/manual/release-notes/rl-2411.section.md:861]: It needs a better rewording.

Comment (ein-shved) [reply nixos/doc/manual/release-notes/rl-2411.section.md:861]: I am not very good in English. Can you suggest something specific, please?

Comment (BonusPlay) [reply nixos/doc/manual/release-notes/rl-2411.section.md:861]:
```suggestion
- `freecad` now supports addons and custom configuration in nix-way, which can be used by calling `freecad.customize`.
```

Comment (ein-shved) [reply nixos/doc/manual/release-notes/rl-2411.section.md]: @BonusPlay thank you!

File: pkgs/by-name/fr/freecad/freecad-utils.nix [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            (lib.lowPrio freecad)
+            bin
+          ];
+        });
+      override = f: makeCustomizable (freecad.override f);
+      overrideAttrs = f: makeCustomizable (freecad.overrideAttrs f);
+    };
+in
+{
+  inherit makeCustomizable;
```
Comment (BonusPlay) [inline pkgs/by-name/fr/freecad/freecad-utils.nix:98]: What's the point of this huge `let ... in` when all it has inside is `makeCustomizable`, that is being returned at the end?

File: pkgs/by-name/fr/freecad/freecad-utils.nix [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+          wrapPathsStr =
+            flag: values:
+            builtins.concatStringsSep " " (
+              builtins.concatMap (p: [
+                "--add-flags"
+                flag
+                "--add-flags"
+                p
+              ]) values
+            );
```
Comment (BonusPlay) [inline pkgs/by-name/fr/freecad/freecad-utils.nix:33]: Pull this out of nested `let .. in`?

File: pkgs/by-name/fr/freecad/freecad-utils.nix [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            let
+              installer = writeShellScript "insteller-${typ}" ''
+                dst="$HOME/.config/FreeCAD/${typ}.cfg"
+                if [ ! -f "$dst" ]; then
+                  mkdir -p "$(dirname "$dst")"
+                  cp --no-preserve=mode,ownership '${val}' "$dst"
+                fi
+              '';
+            in
+            lib.optionalString (val != null) "--run ${installer}";
```
Comment (BonusPlay) [inline pkgs/by-name/fr/freecad/freecad-utils.nix:46]: Pull this out of nested `let .. in`?

File: pkgs/by-name/fr/freecad/freecad-utils.nix [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+          pythonsProcessed = builtins.map (
+            pyt:
+            if builtins.isString pyt then
+              pyt
+            else if builtins.isFunction pyt then
+              "${(python311.withPackages pyt)}/lib/python3.11/site-packages"
+            else
+              throw "Expected string or function as python paths for freecad"
+          ) pythons;
```
Comment (BonusPlay) [inline pkgs/by-name/fr/freecad/freecad-utils.nix:56]: Pull this out of nested `let .. in`?

File: pkgs/by-name/fr/freecad/freecad-utils.nix [hunk 7]
```diff
@@ -0,0 +1,99 @@
+{
+  runCommand,
+  buildEnv,
+  makeWrapper,
+  lib,
+  python311,
```
Comment (BonusPlay) [inline pkgs/by-name/fr/freecad/freecad-utils.nix:6]: Does it have to be specifically 3.11, or could this be just `python3`?

File: pkgs/by-name/fr/freecad/freecad-utils.nix (hunk 3 above)
Comment (ein-shved) [reply pkgs/by-name/fr/freecad/freecad-utils.nix:98]: Suggested by @AndersonTorres and @Pandapip1 [here](https://github.com/NixOS/nixpkgs/pull/347776#discussion_r1811579036)

File: pkgs/by-name/fr/freecad/freecad-utils.nix (hunk 5 above)
Comment (ein-shved) [reply pkgs/by-name/fr/freecad/freecad-utils.nix]: Good point! Thank you!

**Review by BonusPlay — APPROVED** [review]

**Comment by hzeller** [comment]
In the meantime, freecad is at rc4, maybe worthwhile coordinating with #356605 to avoid unnecessary merge conflicts.

**Comment by ein-shved** [comment]
> In the meantime, freecad is at rc4, maybe worthwhile coordinating with #356605 to avoid unnecessary merge conflicts.

How? The only way I know - is to merge one of the patches, then - another. Btw, what have I do next to merge mine?

**Comment by ein-shved** [comment]
@gebner @AndersonTorres @srounce @drupol, hi! Can you merge please this change?

**Comment by ein-shved** [comment]
> I can't.
> 
> Fill a post in this thread:
> 
> https://discourse.nixos.org/t/prs-already-reviewed/2617/2116
> 
> Then wait.

Oh! Did not know. Is this thread specified in docs?

**Comment by AndersonTorres** [comment]
No, it is nowhere specified in docs because it is basically a community effort with no official backing.

**Comment by AndersonTorres** [comment]
I was a committer a long time ago, but I abused the button - by sending a systemd refactor to master branch.

Since then I am in a penance period. I will not ask for commit privileges anytime soon.

**Comment by cab404** [comment]
> I was a committer a long time ago, but I abused the button - by sending a systemd refactor to master branch.
> 
> Since then I am in a penance period. I will not ask for commit privileges anytime soon.

Sounds like a wizard backstory!

Uff... hope you have some sort of a recovery condition set.

**Comment by jmickelin** [comment]
@ein-shved Necroing this to ask, before I create a new issue: I can't get this to work with addons that are packaged as Macros. I tried it with a local macro as follows:

```nix
freecad.customize {
  modules = [
    (lib.fileset.toSource {
      root = ../../freecad-macro;
      fileset = ../../freecad-macro/exportmultiple.FCMacro;
    })
  ];
}
```

I confirmed the wrapper did add the extra module-path argument:

```sh
#! /nix/store/0irlcqx2n3qm6b1pc9rsd2i8qpvcccaj-bash-5.2p37/bin/bash -e
exec -a "$0" "/nix/store/can425gh64qh656q84r3szls4p35l2f5-freecad-1.0.0/bin/FreeCAD"  --module-path /nix/store/960cwkfymn2y4x892pr79fn7mff409p4-source "$@"
```

Yet, FreeCAD does not pick up the macro, either as a user macro or a system macro (which I checked in `Toolbar > Macro > Macros...`).

Checking the existing functionality, FreeCAD does search a specific directory for system-installed macros, as described here: https://wiki.freecad.org/Std_DlgMacroExecute#System_macros

In my tests, using `freecad.customize`, the output of `App.getHomePath()` still points to the Nix store path of the unwrapped FreeCAD program. This means that I can't add the directory myself without having to rebuild the entire program, so I think it could be in need of some similar patching like what you did for modules, to add extra macro paths from the command line.

Since you have already familiarized yourself with the macro loading code, do you figure it would be a lot of work to add a similar patch for macros?

**Comment by ein-shved** [comment]
@jmickelin Yes, I know about macro problems. When I tested, I forgot about macros and perform tests only for entities, distributed by the addon manager. I've fixed (partly) the issue for preference packs (in upstream too), and missed macros. If you will open an issues here and in freecad repo - it will be great. I will handle this in awhile.

**Comment by jmickelin** [comment]
Awesome! I'll get working on the issues, hopefully I can get around to it today or tomorrow.

**Comment by ein-shved** [comment]
> Awesome! I'll get working on the issues, hopefully I can get around to it today or tomorrow.

You can simply workaround it the same way as i did for `userConfig`. Add an extra wrapper argument to launch your script, which checks an existance of your macro at correct path and place it there if one is missing.

**Comment by jmickelin** [comment]
Oh, that's neat! I didn't know `makeWrapper` supported a `--run` argument like that. My current solution is actually to just use Home Manager to write some symlinks into `.local/share/FreeCAD/Macro`. But tweaking the wrapper to create them for you isn't a bad idea.

Another workaround could be to use your `userCfg` parameter, but instead of pointing to a raw XML source file, use  the `substitute` build helper to inject the path of a Nix store path as the value for `ModulePath`:

user.cfg:
```xml
...
        <FCParamGroup Name="Macro">
          <FCText Name="MacroPath">@macroPath@</FCText>
          <FCBool Name="LocalEnvironment" Value="1"/>
          <FCBool Name="RecordGui" Value="1"/>
          <FCBool Name="GuiAsComment" Value="1"/>
          <FCBool Name="ScriptToPyConsole" Value="1"/>
          <FCBool Name="ScriptToFile" Value="0"/>
        </FCParamGroup>
...
```

and then 

```nix
freecad.customize { 
  userCfg = pkgs.substitute { 
    src = ./user.cfg;
    # Assuming this is a derivation built with e.g. buildEnv to combine
    # all the .FCMacros into one directory
    macroPath = myMacroDerivation;
    substitutions = [ 
      "--subst-var"
      "macroPath"
    ];
  };
}
```

What I don't like about User Macros, and the reason I'm hoping to replace them with system-wide ones, is that the config file hardcodes the user's home directory in the `MacroPath`, which has already broken my setup when I migrated between machines. I guess this workaround does solve that issue!

**Pull request merged**